### PR TITLE
Hyperparameter tuning

### LINKED ISSUE #123: Implement hyperparameter search
Integrate an optimizer tool, such as [Ray Tune](https://docs.ray.io/en/latest/tune/index.html), to perform search in the hyperparameter space ("num_folds", "num_epochs", "neurons_per_layer", "num_hidden_layers", "batch_size", "learning_rate", "goal_per_class").

### PR DESCRIPTION
**Pull request by camposandro**
This pull request focuses on model training and hyperparameter tuning for the classifier:

- Integrates Ray Tune to search for the best hyperparameters set.
- Uploads `training_set.csv`, currently used to train the model, to the repo.
- Creates an entrypoint to train and evaluate the model using k-fold cross validation (`train.py`).
- Organizes most MLP related files under the `model` package.

Closes #123.

## Code Quality
- [X] I have read the Contribution Guide
- [x] My code follows the code style of this project
- [x] My code builds (or compiles) cleanly without any errors or warnings
- [x] My code contains relevant comments and necessary documentation

### CODE REVIEW

File: src/superphot_plus/model/classifier.py [hunk 1]
```diff
@@ -314,20 +184,23 @@ def run(
...
             plot_model_metrics(
                 metrics=metrics.get_values(),
                 num_epochs=num_epochs,
-                plot_name=test_names[0],
+                plot_name=run_id,
                 metrics_dir=metrics_dir,
             )
 
+        if self.test_data is None:
+            return best_valid_loss, valid_acc
```
Comment (jeremykubica) [inline src/superphot_plus/model/classifier.py:200]: This is returning only a subset of what the docstring says it will return, which may cause confusion for the users.

File: src/superphot_plus/model/classifier.py [hunk 2]
```diff
@@ -302,6 +171,7 @@ def run(
 
             if val_loss < best_valid_loss:
                 best_valid_loss = val_loss
+                valid_acc = val_acc
                 torch.save(self.state_dict(), model_path)
```
Comment (jeremykubica) [inline src/superphot_plus/model/classifier.py:175]: I know it is not in your code, but it is likely expensive to save each best model to a file. You might want to consider having a copy of the model in memory and only outputting it at the end.

File: src/superphot_plus/model/classifier.py [hunk 3]
```diff
@@ -314,20 +184,23 @@ def run(
...
                 num_epochs=num_epochs,
-                plot_name=test_names[0],
+                plot_name=run_id,
                 metrics_dir=metrics_dir,
             )
 
+        if self.test_data is None:
+            return best_valid_loss, valid_acc
+
+        labels, names, pred_labels, max_probs = self.test(probs_csv_path)
```
Comment (jeremykubica) [inline src/superphot_plus/model/classifier.py:202]: I believe this is all doing the correct thing since test_data = None for the k-fold cross validation, however having the test set here might be error prone for a user.

I suggest breaking this into two functions: training and evaluate on test. Training should take only the training and validation sets. Testing can just use the `test()` function below.

File: src/superphot_plus/model/classifier.py [hunk 4]
```diff
@@ -1,187 +1,52 @@
...
 
     Parameters
     ----------
     config : ModelConfig
         The MLP architecture configuration.
-    data : ModelData
-        The training, test and validation data.
+    train_data : TrainData
+        The training and validation datasets.
+    test_data : TestData, optional
```
Comment (jeremykubica) [inline src/superphot_plus/model/classifier.py:39]: To reduce the probability of a user error, I would suggest removing test_data as a class variable. Instead the classifier should only have access to the training and validation data to start. The `test()` function can take in the test_data explicitly.

See other comments throughout.

File: src/superphot_plus/model/classifier.py [hunk 5]
```diff
@@ -275,22 +146,20 @@ def run(
...
+            valid_dataset,
+        ) = self.train_data
+
+        config_path = os.path.join(models_dir, f"superphot-config-{run_id}.json")
         self.config.save(config_path)
-        model_path = os.path.join(models_dir, f"superphot-model-{test_names[0]}.pt")
 
-        train_iterator = DataLoader(train_data, shuffle=True, batch_size=BATCH_SIZE)
-        valid_iterator = DataLoader(valid_data, batch_size=BATCH_SIZE)
+        model_path = os.path.join(models_dir, f"superphot-model-{run_id}.pt")
```
Comment (jeremykubica) [inline src/superphot_plus/model/classifier.py:156]: Is it worth saving this as a class variable, so you can get it later (to load models for testing)?

File: src/superphot_plus/model/classifier.py [hunk 6]
```diff
@@ -407,26 +280,22 @@ def evaluate_epoch(self, iterator):
 
         return epoch_loss / len(iterator), epoch_acc / len(iterator)
 
-    def test(self, test_features, test_classes, test_names, test_group_idxs, save_path):
+    def test(self, save_path):
```
Comment (jeremykubica) [inline src/superphot_plus/model/classifier.py:283]: As mentioned above, I would suggest having this take in the test data set explicitly.

You could rename `evaluate()` and use this any time you want to evaluate a set of data (test or otherwise).

File: src/superphot_plus/model/classifier.py [hunk 7]
```diff
@@ -407,26 +280,22 @@ def evaluate_epoch(self, iterator):
...
-        test_features : np.ndarray
-            The features array.
-        test_classes : np.ndarray
-            The classes array.
-        test_names : np.ndarray
-            The names array.
-        test_group_idxs : np.ndarray
-            The indices for each test set.
+        save_path
+            Where to store the probability results.
```
Comment (jeremykubica) [inline src/superphot_plus/model/classifier.py:289]: It might be worth having `save_path` default to `None` and skipping the output in that case. This would allow you to skip a file write if you do not need it.

File: src/superphot_plus/model/data.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+class TrainData:
+    """Class that holds the data to train and validate."""
+
+    train_dataset: TensorDataset
+    val_dataset: TensorDataset
+
+    def __iter__(self):
+        return iter(
+            (
+                self.train_dataset,
```
Comment (jeremykubica) [inline src/superphot_plus/model/data.py:18]: I do not understand how the iterator is working here. Does python iterate through the different entries independently? In parallel?

File: src/superphot_plus/tune_model.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+                input_dim=train_features.shape[1],
+                output_dim=output_dim,
+                neurons_per_layer=config["neurons_per_layer"],
+                num_hidden_layers=config["num_hidden_layers"],
+                batch_size=config["batch_size"],
+                learning_rate=config["learning_rate"],
+                normalization_means=mean.tolist(),
+                normalization_stddevs=std.tolist(),
+            ),
+            train_data=TrainData(train_dataset, val_dataset),
```
Comment (jeremykubica) [inline src/superphot_plus/tune_model.py:181]: If you do keep `test_data` as a class variable for `SuperphotClassifier`, I would recommend adding

`test_data=None,  # Only optimize over training and validation`

to make the usage clear to future users.

File: train.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import os
```
Comment (jeremykubica) [inline train.py:1]: Is this meant to be an example script? Or the an implementation that is used to train new models?

If it is an example, maybe put it in an `examples/` subdirectory. Otherwise the parameters should be input arguments.

File: src/superphot_plus/model/classifier.py (hunk 4 above)
Comment (camposandro) [reply src/superphot_plus/model/classifier.py]: Thanks @jeremykubica, I have split this method as you suggested. Turns out this separation helped solve some of other issues you reported in the review :)

File: src/superphot_plus/model/classifier.py (hunk 2 above)
Comment (camposandro) [reply src/superphot_plus/model/classifier.py]: You're right, thanks! Interestingly the impact isn't noticeable, but I decided to apply this change anyway.

File: src/superphot_plus/model/classifier.py (hunk 1 above)
Comment (camposandro) [reply src/superphot_plus/model/classifier.py]: After breaking `run()` into training and testing methods I am now consistently returning the validation loss and accuracy.

File: src/superphot_plus/model/classifier.py (hunk 5 above)
Comment (camposandro) [reply src/superphot_plus/model/classifier.py:149]: I decided to save the best model state as a class variable instead. This way we won't write the whole model to disk and we can get right to testing without too much overhead.

File: train.py (hunk 10 above)
Comment (camposandro) [reply train.py]: Thanks for the feedback! I was thinking to use this script to train new models so I have added the parameters as command line arguments.

File: src/superphot_plus/model/data.py (hunk 8 above)
Comment (camposandro) [reply src/superphot_plus/model/data.py]: I was using this built-in iterator to deconstruct the object as in: `train_dataset, val_dataset = train_data`, and not for iterating over the datasets. Please let me know if this looks like a bad practice and I can revert it!

File: src/superphot_plus/classify_ztf.py [hunk 11]
```diff
@@ -23,180 +30,321 @@
...
+
+        # Initialize output directories
+        os.makedirs(metrics_dir, exist_ok=True)
+        os.makedirs(models_dir, exist_ok=True)
+        os.makedirs(cm_folder, exist_ok=True)
+        os.makedirs(FIT_PLOTS_FOLDER, exist_ok=True)
+
+        # Cleanup previous output content
+        for directory in [metrics_dir, models_dir, cm_folder, FIT_PLOTS_FOLDER]:
+            files = glob.glob(os.path.join(directory, "*"))
```
Comment (delucchi-cmu) [inline src/superphot_plus/classify_ztf.py:106]: Maybe swap these operations? You could use `shutil.rmtree` to remove the directory and all contents recursively, then recreate the directory.

File: src/superphot_plus/classify_ztf.py [hunk 12]
```diff
@@ -23,180 +30,321 @@
...
-    Parameters
-    ----------
-    features_orig : np.ndarray
-        Array of fit features of all samples.
 
-    Returns
-    ---------
-    features : np.ndarray
-        Array of adjusted fit features.
+# pylint: disable=too-many-instance-attributes)
```
Comment (delucchi-cmu) [inline src/superphot_plus/classify_ztf.py:47]:
```suggestion
# pylint: disable=too-many-instance-attributes
```

File: src/superphot_plus/classify_ztf.py [hunk 13]
```diff
@@ -23,180 +30,321 @@
...
+        )
 
-        train_labels = labels[train_index]
-        val_labels = labels[val_index]
+    def generate_train_data(self, names, labels, redshifts, train_indices):
+        """Creates training and validation data, oversampling when needed.
+
+        Parameters
+        ----------
+        names The full list of ZTF objects.
```
Comment (delucchi-cmu) [inline src/superphot_plus/classify_ztf.py:262]: These param lists don't quite meet the numpy format. See rendered param blocks in https://superphot-plus--124.org.readthedocs.build/en/124/autoapi/src/superphot_plus/classify_ztf/index.html#src.superphot_plus.classify_ztf.CrossValidationTrainer.generate_test_data

**Review by jeremykubica — CHANGES_REQUESTED** [review]

File: src/superphot_plus/utils.py [hunk 14]
```diff
@@ -325,6 +326,9 @@ def calculate_neg_chi_squareds(cubes, t, f, ferr, b, ordered_bands=["r", "g"], r
     log_likelihoods : np.ndarray
         The log likelihoods for each object.
     """
+    if ordered_bands is None:
```
[the remark this answers is not in the document]
Comment (delucchi-cmu) [reply src/superphot_plus/utils.py:329]: It's a pylint warning to have a default value as a list. The behavior is identical.

File: src/superphot_plus/classify_ztf.py (hunk 13 above)
Comment (camposandro) [reply src/superphot_plus/classify_ztf.py]: Thanks! I am used to writing docstrings using the numpy notatio, but for some reason I missed these.

File: src/superphot_plus/classify_ztf.py [hunk 15]
```diff
@@ -23,180 +30,321 @@
...
-    ----------
-    features_orig : np.ndarray
-        Array of fit features of all samples.
 
-    Returns
-    ---------
-    features : np.ndarray
-        Array of adjusted fit features.
+# pylint: disable=too-many-instance-attributes)
+class CrossValidationTrainer:
```
[the remark this answers is not in the document]
Comment (camposandro) [reply src/superphot_plus/classify_ztf.py:48]: Correct! I'm only performing parameter tuning in `tune.py`, where I am holding out a test set. Here the CrossValidationTrainer wraps the code that trains the model and performs validation. Let's further discuss with @kdesoto-astro if there's any reason behind the usage of the test set on cross validation here.

File: src/superphot_plus/model/config.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+            "normalization_stddevs": self.normalization_stddevs,
+            "batch_size": self.batch_size,
+            "learning_rate": self.learning_rate,
+        }
+        with open(filename, "w", encoding="utf-8") as f:
+            json.dump(data_dict, f)
+
+    @classmethod
+    def load(cls, filename):
+        """Load configuration data from a JSON file."""
```
[the remark this answers is not in the document]
Comment (camposandro) [reply src/superphot_plus/model/config.py]: Sure, to keep it consistent for now I have changed these configuration files to the YAML format!

**Review by kdesoto-astro — APPROVED** [review]
Looks great!
